\chapter{Sample CLI Script Generated by Provenance Replay}
\label{app:CLIScript}

\begin{footnotesize}
\begin{verbatim}
#!/usr/bin/env bash
###############################################################################
# Auto-generated by provenance_lib v.0.0.1 at 12:09:59 AM on 11 Apr, 2022

# For User Support, post to the Community Plugin Support channel of the QIIME 2
# Forum: https://forum.qiime2.org
# Documentation/issues: https://github.com/ChrisKeefe/provenance_py.git

# UUIDs of all target QIIME 2 Results are shown at the end of the file

# Instructions for use:
# 1. Open this script in a text editor or IDE. Support for BASH
#    syntax highlighting can be helpful.
# 2. Search or scan visually for '<' or '>' characters to find places where
#    user input (e.g. a filepath or column name) is required. These must be
#    replaced with your own values. E.g. <column name> -> 'patient_id'.
#    Failure to remove '<' or '>' may result in `No such File ...` errors
# 3. Search for 'TODO' comments in the script, and respond as directed.
# 4. Remove all 'TODO' comments from the script completely. Failure to do so
#    may result in 'Missing Option' errors
# 5. Adjust the arguments to the commands below to suit your data and metadata.
#    If your data is not identical to that in the replayed analysis,
#    changes may be required. (e.g. sample ids or rarefaction depth)
# 6. Optional: replace any filenames in this script that begin with 'XX' with
#    unique file names to ensure they are preserved. QIIME 2 saves all outputs
#    from all actions in this script to disk regardless of whether those
#    outputs were in the original collection of replayed results. The filenames
#    of "un-replayed" artifacts are prefixed with 'XX' so they may be easily
#    located. These names are not guaranteed to be unique, so 'XX_table.qza'
#    may be overwritten by another 'XX_table.qza' later in the script.
# 7. Activate your replay conda environment, and confirm you have installed all
#    plugins used by the script.
# 8. Run this script with `bash <path to this script>`, or copy-paste commands
#    into the terminal for a more interactive analysis.
# 9. Optional: to delete all results not required to produce the figures and
#    data used to generate this script, navigate to the directory in which you
#    ran the script and `rm XX*.qz*`
###############################################################################

# This tells bash to -e exit immediately if a command fails
# and -x show all commands in stdout so you can track progress
set -e -x

qiime tools import \
  --type 'EMPSingleEndSequences' \
  --input-path <your data here> \
  --output-path emp-single-end-sequences-0.qza

# Replay attempts to represent metadata inputs accurately, but metadata .tsv
# files are merged automatically by some interfaces, rendering distinctions
# between file inputs invisible in provenance. We output the recorded
# metadata to disk to enable visual inspection.

# The following command may have received additional metadata .tsv files. To
# confirm you have covered your metadata needs adequately, review the
# original metadata, saved at 'recorded_metadata/demux_emp_single_0/'

qiime demux emp-single \
  --i-seqs emp-single-end-sequences-0.qza \
  --m-barcodes-file <barcodes-0.tsv> \
  --m-barcodes-column <column name> \
  --p-no-rev-comp-barcodes \
  --p-no-rev-comp-mapping-barcodes \
  --o-per-sample-sequences per-sample-sequences-0.qza \
  --o-error-correction-details XX_error_correction_details

qiime dada2 denoise-single \
  --i-demultiplexed-seqs per-sample-sequences-0.qza \
  --p-trunc-len 120 \
  --p-trim-left 0 \
  --p-max-ee 2.0 \
  --p-trunc-q 2 \
  --p-chimera-method consensus \
  --p-min-fold-parent-over-abundance 1.0 \
  --p-n-threads 1 \
  --p-n-reads-learn 1000000 \
  --p-hashed-feature-ids \
  --o-representative-sequences representative-sequences-0.qza \
  --o-table table-0.qza \
  --o-denoising-stats XX_denoising_stats

qiime phylogeny align-to-tree-mafft-fasttree \
  --i-sequences representative-sequences-0.qza \
  --p-n-threads 1 \
  --p-mask-max-gap-frequency 1.0 \
  --p-mask-min-conservation 0.4 \
  --o-rooted-tree rooted-tree-0.qza \
  --o-alignment XX_alignment \
  --o-masked-alignment XX_masked_alignment \
  --o-tree XX_tree

# The following command may have received additional metadata .tsv files. To
# confirm you have covered your metadata needs adequately, review the
# original metadata, saved at
# 'recorded_metadata/diversity_core_metrics_phylogenetic_0/'

qiime diversity core-metrics-phylogenetic \
  --i-table table-0.qza \
  --i-phylogeny rooted-tree-0.qza \
  --p-sampling-depth 1109 \
  --m-metadata-file <metadata-0.tsv> \
  # TODO: The following parameter name was not found in your current
  # QIIME 2 environment. This may occur when the plugin version you have
  # installed does not match the version used in the original analysis.
  # Please see the docs and correct the parameter name before running.
  --?-n-jobs 1 \
  --o-unweighted-unifrac-emperor unweighted-unifrac-emperor-0.qzv \
  --o-rarefied-table XX_rarefied_table \
  --o-faith-pd-vector XX_faith_pd_vector \
  --o-observed-features-vector XX_observed_features_vector \
  --o-shannon-vector XX_shannon_vector \
  --o-evenness-vector XX_evenness_vector \
  --o-unweighted-unifrac-distance-matrix XX_unweighted_unifrac_distance_matrix \
  --o-weighted-unifrac-distance-matrix XX_weighted_unifrac_distance_matrix \
  --o-jaccard-distance-matrix XX_jaccard_distance_matrix \
  --o-bray-curtis-distance-matrix XX_bray_curtis_distance_matrix \
  --o-unweighted-unifrac-pcoa-results XX_unweighted_unifrac_pcoa_results \
  --o-weighted-unifrac-pcoa-results XX_weighted_unifrac_pcoa_results \
  --o-jaccard-pcoa-results XX_jaccard_pcoa_results \
  --o-bray-curtis-pcoa-results XX_bray_curtis_pcoa_results \
  --o-weighted-unifrac-emperor XX_weighted_unifrac_emperor \
  --o-jaccard-emperor XX_jaccard_emperor \
  --o-bray-curtis-emperor XX_bray_curtis_emperor


###############################################################################
# The following QIIME 2 Results were parsed to produce this script:
# ffb7cee3-2f1f-4988-90cc-efd5184ef003
###############################################################################
\end{verbatim}
\end{footnotesize}
